### PR TITLE
ovsx-client: support multiple registries

### PR DESCRIPTION
**Pull request by paul-marechal**
Refactors the `OVSXClient` into a simple interface that can easily be implemented by components such as `OVSXHttpClient` or others.

Add a new `OVSXRouterClient` that may route OpenVSX queries to different registries based on an `OVSXRouteConfig` configuration. The path to the configuration file can be specified when starting the backend using the `--ovsx-router-config` CLI param. When this configuration is not specified, we default to the previously configured OpenVSX registry.

The goal of these changes was to minimally impact Theia's plugin code base by simply replacing the client interfaces we use to communicate with an OpenVSX registry. This proved more difficult than expected, especially when merging search results: pagination may currently be broken when dealing with multiple registries.

Add a sample mock OpenVSX server running as part of Theia's backend. You can find the plugins served by this mock server under `examples/api-samples/plugins-db`. This mock is there to allow easier testing of the multi-registry feature.

#### How to test

- Starting the backend without the `--ovsx-router-config` CLI argument should just query `open-vsx.org` for plugins.
- Search extensions for `sample-namespace`: It should only emit a request to the mock OpenVSX server and show its results
- Search for `sample-namespac`: It should emit the request to both (mock, openvsx) registries.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by paul-marechal** [comment]
While there's benefits to having our main OpenVSX client do the routing between multiple registries, the "biggest" problem ended up to find a way to properly merge the search results, as the order somewhat matters. In its current shape this PR doesn't do a great merging job: I just interleave results from each registries. In most situations this is less than perfect, but I'm not sure what we should do.

Surprisingly, it doesn't seem to be that big of a deal in practice. Using the mock OpenVSX server it seems relatively easy to list plugins from both registries, and the more specific your search term is and the more relevant the list seems to become.

Tell me what you think of the current state of this feature, and please suggest improvements if you have any!

**Comment by paul-marechal** [comment]
While this could serve us for now, I think Theia's plugin system needs to be refactored in a more plugin-type-agnostic way while supporting multiple plugin providers. The extension view would list extensions based on the provider, as this will avoid the issues I had here when trying to find a way to merge everything into a single meaningful list.

**Comment by msujew** [comment]
Note to self (and Paul): I'm looking at this today 👍

**Review by msujew — COMMENTED** [review]
The change looks mostly good to me. What I couldn't figure out was whether the merging of results deals with duplicate entries (so multiple registries yielding the same extension) as expected. Would the extension view should them as duplicates or is only one shown?

**Comment by MatthewKhouzam** [comment]
I have tested this, it brings value to Theia. The order of plug-ins is not logical, but this is an issue that I can live with. Patches are welcome on that.

**Comment by msujew** [comment]
I'm fine with the change, but I still have an [open question](https://github.com/eclipse-theia/theia/pull/12040#pullrequestreview-1331053865).

**Comment by paul-marechal** [comment]
@msujew Application makers have to make sure to write their configuration in such a way that there are no duplicates returned by the router client. I don't know what the `@theia/vsx-registry` would do if the client were to return duplicates.

**Comment by MatthewKhouzam** [comment]
I am dragging @ghuntley in... This is something that may be of interest for https://github.com/coder/code-marketplace

File: dev-packages/cli/src/theia.ts [hunk 1]
```diff
@@ -355,17 +357,44 @@ async function theiaCli(): Promise<void> {
                 'proxy-url': {
                     describe: 'Proxy URL'
                 },
-                'proxy-authentification': {
+                'proxy-authorization': {
                     describe: 'Proxy authentification information'
```
Comment (kkistm) [inline dev-packages/cli/src/theia.ts:361]: Looks like `describe` doesn't correspond to the object name anymore.

**Comment by kkistm** [comment]
About presenting the results of search from several repositories, why not to add a feature to a repository to expose some metadata about itself? For example, a display name and icon. Then the extension list could show an icon of the repository it originates from as a decoration in some corner or something. A sorting of the results according to the repository of the origin could also be possible. OpenVSX-style repository could implement it as `/about` endpoint or something like this.
What do you think about it?

File: packages/vsx-registry/src/node/vsx-extension-resolver.ts [hunk 2]
```diff
@@ -49,10 +50,12 @@ export class VSXExtensionResolver implements PluginDeployerResolver {
         const client = await this.clientProvider();
         if (options) {
             console.log(`[${id}]: trying to resolve version ${options.version}...`);
-            extension = await client.getExtension(id, { extensionVersion: options.version, includeAllVersions: true });
+            const { extensions } = await client.query({ extensionId: id, extensionVersion: options.version, includeAllVersions: true });
+            extension = extensions[0];
```
[the remark this answers is not in the document]
Comment (paul-marechal) [reply packages/vsx-registry/src/node/vsx-extension-resolver.ts:54]: Then `extension` will be `undefined` which is accounted for later.

File: packages/vsx-registry/src/node/vsx-cli.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import { OVSXRouterConfig } from '@theia/ovsx-client';
+import * as fs from 'fs';
+
+@injectable()
+export class VsxCli implements CliContribution {
+
+    ovsxRouterConfig: OVSXRouterConfig | undefined;
+
+    configure(conf: Argv<{}>): void {
+        conf.option('ovsx-router-config', { description: 'JSON configuration file for the OVSX router client', type: 'string' });
```
[the remark this answers is not in the document]
Comment (paul-marechal) [reply packages/vsx-registry/src/node/vsx-cli.ts:29]: Yeah I wish we abstracted away the way to configure such things so that 1) we don't directly depend on the yargs APIs and 2) when defining a CLI option it would also look at the environment automatically.

I can see to align both options.

File: packages/vsx-registry/src/browser/vsx-extensions-contribution.ts [hunk 4]
```diff
@@ -48,7 +48,8 @@ export class VSXExtensionsContribution extends AbstractViewContribution<VSXExten
     @inject(LabelProvider) protected readonly labelProvider: LabelProvider;
     @inject(ClipboardService) protected readonly clipboardService: ClipboardService;
     @inject(PreferenceService) protected readonly preferenceService: PreferenceService;
-    @inject(OVSXClientProvider) protected readonly clientProvider: OVSXClientProvider;
+    @inject(OVSXClientProvider) protected clientProvider: OVSXClientProvider;
```
[the remark this answers is not in the document]
Comment (paul-marechal) [reply packages/vsx-registry/src/browser/vsx-extensions-contribution.ts:51]: I somewhat accidentally removed the `readonly` modifier as I was dealing with these fields.

FWIW I think it's a mistake to make injected fields readonly. I don't see how it prevents any kind of issue. It actually prevents manual field injection when testing as we cannot assign values to the `readonly` fields.

I'll remove the `readonly` modifiers on injected fields in this class to make it more consistent then.

File: dev-packages/ovsx-client/src/ovsx-types.ts [hunk 5]
```diff
@@ -56,14 +105,35 @@ export interface VSXSearchParam {
...
-    readonly extensions: VSXSearchEntry[];
+    error?: string;
+    offset: number;
+    extensions: VSXSearchEntry[];
+}
+
+/** @deprecated since 1.31.0 use {@link VSXQueryOptions} instead */
+export type VSXQueryParam = VSXQueryOptions;
+export interface VSXQueryOptions {
+    namespaceName?: string;
```
[the remark this answers is not in the document]
Comment (paul-marechal) [reply dev-packages/ovsx-client/src/ovsx-types.ts:116]: This type should reflect what the OpenVSX API expects, I'll add a link to that.

File: dev-packages/ovsx-client/src/ovsx-types.ts [hunk 6]
```diff
@@ -14,14 +14,58 @@
 // SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 // *****************************************************************************
 
+import { MaybePromise } from './types';
+
+export interface ExtensionLike {
+    name: string
```
[the remark this answers is not in the document]
Comment (paul-marechal) [reply dev-packages/ovsx-client/src/ovsx-types.ts:20]: The linter currently doesn't care about semi colons in interfaces for some reason, which makes it difficult to be consistent... I'll see to add the missing ones.

File: dev-packages/ovsx-client/src/ovsx-types.ts [hunk 7]
```diff
@@ -14,14 +14,58 @@
...
+        const [namespace, name] = left.split('.', 2);
+        return {
+            name,
+            namespace,
+            version
+        };
+    }
+}
+
+export interface OVSXClient {
```
[the remark this answers is not in the document]
Comment (paul-marechal) [reply dev-packages/ovsx-client/src/ovsx-types.ts:46]: I get what you mean about DTOs but I felt like this interface was generic enough to be defined in what seems to be a file containing all types used by this package. Implementations live in their own separate files.

File: packages/vsx-registry/src/browser/vsx-extensions-contribution.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (kkistm) [reply packages/vsx-registry/src/browser/vsx-extensions-contribution.ts:50]: As an old server-side developer I would prefer to have `readonly` everywhere it could be placed :) I would propose to do vice versa, to add `readonly` and configure the tests via tiny Inversify snippets. What do you think about it?

[the remark this answers is not in the document]
Comment (paul-marechal) [reply packages/vsx-registry/src/browser/vsx-extensions-contribution.ts:50]: To add to my case, consider the following [snippet](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgLIE8CSBbbBXMOAIwBsIBlaAN2CWQG8AoZF5KCOAEwHsQT1kMYBBKcAggC5k3IgCsICMM1bsuvfoOGiAQlJnzFylgGdu2CKghgAFt04AKAJRSq3YJyPSb0SzbtOXNw8AX0ZGTgUSOHZkBF5jMGRjaloIKQwcfEJSChSkMOSoGiQAOiERcWQAXgZg5AB6euRoKG4oZDhjZoAPAAcFSA9C4ogS03NfW05q5CdqgD5ahqaAFXJkYC6hEBRraAgAQkYgA):

```ts
interface MyImmutableService {
    readonly fieldA: object
    readonly fieldB: object
    someMethod(): void
    otherMethod(): void
}

declare const service: MyImmutableService

service.fieldA = {} // error as expected
service.someMethod = () => {} // TS is fine here!
```

If you want the methods to be readonly too you need to write things like:

```ts
interface MyImmutableService {
    readonly fieldA: object
    readonly fieldB: object
    readonly someMethod: () => void
    readonly otherMethod: () => void
}
```

But at this point you really denature service interface definition...

Long story short: I don't think we should care about making `readonly` fields in services while it may still make sense for "data-objects" that may be returned and passed around by APIs.

[the remark this answers is not in the document]
Comment (kkistm) [reply packages/vsx-registry/src/browser/vsx-extensions-contribution.ts:50]: I wrote about server-side for exact reasons you mentioned above: more immutability is better. If something could be made immutable, or at least declared immutable, I consider it is good. `readonly`, as far as I understand, should prevent somebody from assigning a value to a field accidentally in the code. And also it says to the reader "this field is not supposed to be changed after it is assigned". But please take all my arguments with a grain of salt, my Theia and TS experience should be considered as basic at this stage.
And just as a side note, for me the ability to create an instance of interface and start assigning its fields is quite crazy :) I understand why it works like this, but still...
As for `readonly` for DI fields, I would propose to make it a bit wider topic than a particular review. If there are any coding guidelines for Theia, what do they say about it? If nothing, I would propose to decide in a group (next meeting, maybe?) and then make it a guideline. And then write new code and rework old one, when there is an opportunity, according to the guideline. What do you think about it?

File: packages/vsx-registry/src/browser/vsx-extensions-model.ts [hunk 8]
```diff
@@ -59,20 +74,71 @@ export class VSXExtensionsModel {
...
-        await Promise.all([
+        this.initialized = Promise.all([
             this.initInstalled(),
             this.initSearchResult(),
             this.initRecommended(),
-        ]);
-        this.initialized.resolve();
+        ]).then(() => { }, error => {
+            console.error(error);
+        });
```
Comment (msujew) [inline packages/vsx-registry/src/browser/vsx-extensions-model.ts:91]: Minor: Should `catch` be more straightforward here?
```suggestion
        ]).catch(error => {
            console.error(error);
        });
```

**Review by msujew — APPROVED** [review]
Alright, looks good to me 👍 Don't forget to rebase before merging :)
